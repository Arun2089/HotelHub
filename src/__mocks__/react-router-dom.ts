const mock = jest.requireActual("react-router-dom");

module.exports = {
  ...mock,
  useNavigate: jest.fn(),
  useParams: jest.fn(),
  useLocation: jest.fn(),
  useHistory: jest.fn(),
};
