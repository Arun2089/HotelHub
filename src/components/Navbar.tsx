import { useState } from "react";
import {
  Box,
  Typography,
  Menu,
  Tooltip,
  Avatar,
  MenuItem,
  Toolbar,
  IconButton,
} from "@mui/material";
import SegmentIcon from "@mui/icons-material/Segment";

import { SETTINGS } from "@constant/index";

type navbarPropsTypes = {
  toogleSideBar: () => void;
};

const Navbar = ({ toogleSideBar }: navbarPropsTypes) => {
  const [anchorElUser, setAnchorElUser] = useState<null | HTMLElement>(null);

  const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>): void => {
    setAnchorElUser(event.currentTarget);
  };

  const handleCloseUserMenu = (): void => {
    setAnchorElUser(null);
  };

  const positioning: {
    vertical: number | "top" | "center" | "bottom";
    horizontal: number | "center" | "right" | "left";
  } = {
    vertical: "top",
    horizontal: "right",
  };
  return (
    <Box className="h-20 bg-background-default" position="static">
      <Toolbar className="h-20">
        <IconButton
          className="mr-4"
          size="large"
          edge="start"
          color="inherit"
          aria-label="menu"
          onClick={toogleSideBar}
        >
          <SegmentIcon className="text-3xl" />
        </IconButton>
        <Typography className="grow font-extrabold text-3xl" variant="h5">
          DashBoard
        </Typography>
        <Box className="grow-0">
          <Tooltip title="Open settings">
            <IconButton className="p-0" onClick={handleOpenUserMenu}>
              <Avatar className="bg-primary-main" alt="Profile Logo" />
            </IconButton>
          </Tooltip>
          <Menu
            className="mt-11"
            id="menu-appbar"
            anchorEl={anchorElUser}
            anchorOrigin={positioning}
            keepMounted
            transformOrigin={positioning}
            open={Boolean(anchorElUser)}
            onClose={handleCloseUserMenu}
          >
            {SETTINGS.map((setting: string) => (
              <MenuItem key={setting}>
                <Typography textAlign="center">{setting}</Typography>
              </MenuItem>
            ))}
          </Menu>
        </Box>
      </Toolbar>
    </Box>
  );
};

export default Navbar;
