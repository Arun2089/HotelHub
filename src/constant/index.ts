import {
  Dashboard as DashboardIcon,
  AccountBox as AccountBoxIcon,
  Assessment as AssessmentIcon,
  RestaurantMenu as RestaurantMenuIcon,
  BedroomParent as BedroomParentIcon,
  AssignmentIndRounded as AssignmentIndRoundedIcon,
} from "@mui/icons-material";

import { SidebarValueType, AdminDashboardDetailsTypes } from "@ts/types";

export const ADMIN_SIDEBAR: SidebarValueType[] = [
  {
    field: "Dashboard",
    icon: DashboardIcon,
    path: "dashboard",
  },
  {
    field: "Student Info.",
    icon: AccountBoxIcon,
    path: "students",
  },
  {
    field: "Complaints",
    icon: AssessmentIcon,
    path: "complaints",
  },
  {
    field: "Canteen Menu",
    icon: RestaurantMenuIcon,
    path: "canteenMenu",
  },
  {
    field: "Rooms",
    icon: BedroomParentIcon,
    path: "rooms",
  },
];

export const NAVBAR_TITLE: { [key: string]: string } = {
  dashboard: "Dashboard",
  students: "Student Info.",
  complaints: "Complaints",
  canteenMenu: "Canteen Menu",
  rooms: "Rooms",
};

export const STUDENT_SIDEBAR: SidebarValueType[] = [
  {
    field: "Dashboard",
    icon: DashboardIcon,
    path: "dashboard",
  },
  {
    field: "Complaints",
    icon: AssessmentIcon,
    path: "complaints",
  },
  {
    field: "Canteen Menu",
    icon: RestaurantMenuIcon,
    path: "conteenMenu",
  },
];

export const SETTINGS: string[] = ["Profile", "Logout"];

export const DASHBOARD_ENDPOINT: string = "dashboard";

export const USER_DATA_URL: string =
  "https://api.npoint.io/de4471c76c2129b205d7";

export const ADMIN_DASHBOARD_DETAIL_URL: string =
  "https://api.npoint.io/afae6426adf49484df48";

export const NOTICES_URL: string = "http://localhost:3001/notice";

export const ROOM_STATUS_DATA_URL: string =
  "https://api.npoint.io/89bedc5ce0a4a2482847";

export const STAFF_LIST_URL = "https://api.npoint.io/a3ada2b77246b3d38e20";

export const COMPLAINTS_STATS_URL =
  "https://api.npoint.io/34d932b8609356c8bd40";

export const STUDENT_INFO_URL = "http://localhost:3001/studentInfo";

export const COMPLAINTS_URL = "http://localhost:3001/complaints";

export const ADMIN_DASHBOARD_DETAIL: AdminDashboardDetailsTypes[] = [
  {
    label: "Students",
    icon: AccountBoxIcon,
    field: "numberOfStudents",
    color: "bg-primary-main",
    path: "students",
  },
  {
    label: "Complaints",
    icon: AssessmentIcon,
    field: "complaitsPending",
    color: "bg-common-lightBlue",
    path: "complaints",
  },

  {
    label: "Rooms",
    icon: BedroomParentIcon,
    field: "numberOfRooms",
    color: "bg-success-light",
    path: "rooms",
  },
  {
    label: "Staff",
    icon: AssignmentIndRoundedIcon,
    field: "numberOfStaff",
    color: "bg-secondary-main",
    path: "",
  },
];

export const STUDENT_NAME_INPUT = {
  label: "Student Name",
  name: "studentName",
  type: "text",
};

export const STUDENT_EMAIL_INPUT = {
  label: "Student Email",
  name: "email",
  type: "email",
};

export const STUDENT_PHONE_INPUT = {
  label: "Mobile Number",
  name: "mobileNumber",
  type: "tel",
};

export const GUARDIAN_NAME_INPUT = {
  label: "Guardian Name",
  name: "guardianName",
  type: "text",
};

export const GUARDIAN_PHONE_INPUT = {
  label: "Guardian Phone no.",
  name: "guardianPhoneNumber",
  type: "tel",
};

export const ADDRESS_INPUT = {
  label: "Address",
  name: "address",
  type: "text",
};

export const PASSWORD_INPUT = {
  label: "Password",
  name: "password",
  type: "password",
};

export const CONFIRM_PASSWORD_INPUT = {
  label: "Confirm Password",
  name: "confirmPassword",
  type: "password",
};

export const READ_ONLY_SX_VALUES = {
  "& .MuiOutlinedInput-root": {
    "& fieldset": {
      borderColor: "inherit",
    },
    "&.Mui-focused fieldset": {
      borderColor: "inherit",
      borderWidth: 1,
    },
    "&:hover fieldset": {
      borderColor: "inherit",
    },
  },
};

export const COMPLAINT_STATUS = ["all", "pending", "resolved", "invalid"];

export const ID_TYPE = ["Aadhar Card", "Pan Card", "Institute Card"];

export const STEPPER_FORM_STEPS_NAME = [
  "Personal Info",
  "Guardian Info",
  "Password",
  "Goverment Id.",
];
