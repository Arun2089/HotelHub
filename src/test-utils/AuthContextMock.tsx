import { AuthContext, AuthContextType } from "@src/context/AuthContext";

jest.mock("@src/utils/index.ts", () => ({
  fetchData: jest.fn().mockResolvedValue({
    id: 1,
    email: "johnDoe@gmail.com",
    role: { type: "admin" },
    username: "John Doe",
  }),
}));

type PropsWithChildren = {
  children: React.ReactNode;
  value: Partial<AuthContextType>;
};

const defaultValues = {
  user: {
    id: 1,
    email: "johnDoe@gmail.com",
    role: "admin",
    name: "John Doe",
  },
  updateUserDetails: jest.fn(),
  getUser: jest.fn().mockResolvedValue(true),
  isLoading: false,
  handleLogout: jest.fn(),
  handleLogin: jest.fn(),
  handleGoogleLogin: jest.fn(),
};

const AuthProviderMock = ({ children, value }: PropsWithChildren) => {
  return (
    <AuthContext.Provider value={{ ...defaultValues, ...value }}>
      {children}
    </AuthContext.Provider>
  );
};
export default AuthProviderMock;
