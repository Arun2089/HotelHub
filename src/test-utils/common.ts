export const setTimeOutMock = (ms: number) =>
  new Promise((resolve) => setTimeout(resolve, ms));
