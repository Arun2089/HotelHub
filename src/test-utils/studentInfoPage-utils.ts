import { fireEvent } from "@testing-library/react";

export const mockStudent = () => {
  return {
    studentName: "",
    roomNumber: null,
    mobileNumber: "",
    guardianName: "",
    guardianPhoneNumber: "",
    address: "",
    email: "",
    password: "",
    confirmPassword: "",
    govId: "",
    govIdImage: "",
  };
};

export const mockHandleChange = jest.fn();
export const mockValidateError = jest.fn();
export const mockGetHelperText = jest.fn();
export const mockHandleClick = jest.fn();

export function changeMultipleFields(
  textFields: HTMLElement[],
  details: { [key: string]: string | number | null | undefined }
) {
  textFields.forEach((textField) => {
    const name = textField.getAttribute("name");
    if (name) {
      fireEvent.change(textField, {
        target: { value: details[name] },
      });
    }
  });
}

export const studentDetails: {
  [key: string]: string | number | null | undefined;
} = {
  studentName: "John Doe",
  email: "john@gmail.com",
  mobileNumber: "9994567890",
};

export const guardianDetials: {
  [key: string]: string | number | null | undefined;
} = {
  guardianName: "John Doe",
  guardianPhoneNumber: "9994567890",
  address: "123, Main Street, New York",
};

export const passwordDetails: {
  [key: string]: string | number | null | undefined;
} = {
  password: "12345678",
  confirmPassword: "12345678",
};

export const govermentidDetails: {
  [key: string]: string | number | null | undefined;
} = {
  govId: "Pan Card",
  govIdImage: "Base64 string",
};
