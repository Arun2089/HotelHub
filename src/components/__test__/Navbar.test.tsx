import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { useLocation, useNavigate } from "react-router-dom";
import Navbar from "../Navbar";
import AuthProviderMock from "@src/test-utils/AuthContextMock";
import { NAVBAR_TITLE_ADMIN, NAVBAR_TITLE_STUDENT } from "@src/constant";
import { act } from "react-dom/test-utils";

const toogleSideBarMock = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
  useNavigate: jest.fn(),
}));

const contextValues = (role: string) => {
  return {
    user: {
      id: 1,
      email: "johnDoe@gmail.com",
      role: role,
      name: "John Doe",
    },
  };
};

function setupMocks(pathname: string) {
  (useLocation as jest.Mock).mockReturnValue({ pathname: `/${pathname}` });
  (useNavigate as jest.Mock).mockReturnValue(jest.fn());
}

describe("Navbar", () => {
  it("Snapshot testing", () => {
    setupMocks("dashboard");
    const { asFragment } = render(
      <AuthProviderMock value={contextValues("admin")}>
        <Navbar toogleSideBar={toogleSideBarMock} />
      </AuthProviderMock>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the correct title for an admin user", () => {
    setupMocks("dashboard");
    render(
      <AuthProviderMock value={contextValues("admin")}>
        <Navbar toogleSideBar={toogleSideBarMock} />
      </AuthProviderMock>
    );
    expect(
      screen.getByText(NAVBAR_TITLE_ADMIN["dashboard"])
    ).toBeInTheDocument();
  });

  it("renders the correct title for a student user", () => {
    setupMocks("dashboard");
    render(
      <AuthProviderMock value={contextValues("student")}>
        <Navbar toogleSideBar={toogleSideBarMock} />
      </AuthProviderMock>
    );
    expect(
      screen.getByText(NAVBAR_TITLE_STUDENT["dashboard"])
    ).toBeInTheDocument();
  });

  it("open the user menu when the user avatar is clicked", () => {
    setupMocks("dashboard");
    render(
      <AuthProviderMock value={contextValues("student")}>
        <Navbar toogleSideBar={toogleSideBarMock} />
      </AuthProviderMock>
    );
    const userAvatar = screen.getByRole("button", { name: "Open settings" });
    act(() => {
      userAvatar.click();
    });
    expect(screen.getByText("Logout")).toBeInTheDocument();
  });
});
