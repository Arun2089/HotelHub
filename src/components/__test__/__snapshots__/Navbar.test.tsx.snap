// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="h-20 bg-background-default MuiBox-root css-1vu2yqv"
    data-testid="navbar"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular h-20 px-2.5 md:px-5 animate-slideDown css-hyum1k-MuiToolbar-root"
    >
      <button
        aria-label="menu"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge mr-4 css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium text-3xl css-i4bv87-MuiSvgIcon-root"
          data-testid="SegmentIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 18h12v-2H9v2zM3 6v2h18V6H3zm6 7h12v-2H9v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <h5
        class="MuiTypography-root MuiTypography-h5 grow font-bold text-2xl md:text-3xl css-ag7rrr-MuiTypography-root"
      >
        Dashboard
      </h5>
      <div
        class="grow-0 MuiBox-root css-0"
      >
        <button
          aria-label="Open settings"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium p-0 css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault bg-primary-main w-8 h-8 md:h-10 md:w-10 css-2s90m6-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
              data-testid="PersonIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
